flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobGraph.java

```diff
@@ -581,6 +582,8 @@ public String toString() {
 	}
 
 	public void uploadUserArtifacts(InetSocketAddress blobServerAddress, Configuration clientConfig) throws IOException {
+		zipUserArtifacts();
```

Comment:
Could we say that all `userArtifacts` are already zipped if they are a directory before being added via `addUserArtifact`. This means that the caller is responsible for the zipping. That way we could get rid of modifying the `JobGraph` as a side effect of `uploadUserArtifacts`.

Reply:
We would have to do this either right away in `ExEnv#registerCachedFile` or in `ExEnv#registerCachedFilesWithPlan`.

Reply:
Or in `JobGraphGenerator` and `StreamingJobGraphGenerator`.

Reply:
I don't know that part of the code, my guess was that we want to keep the transition from `Plan -> JobGraph` straight-forward.